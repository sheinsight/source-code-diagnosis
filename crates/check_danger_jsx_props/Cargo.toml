[package]
edition = "2024"
name    = "check_danger_jsx_props"
version = "0.1.0"

[dependencies]
anyhow      = { workspace = true }
beans       = { path = "../beans" }
camino      = { workspace = true }
insta       = "1.36.1"
napi        = { workspace = true }
napi-derive = { workspace = true }
pathdiff    = { workspace = true }
rayon       = { workspace = true }
ropey       = "1.0.0"
serde       = { workspace = true }
serde_json  = { workspace = true }
utils       = { path = "../utils" }
wax         = { workspace = true }

oxc        = { workspace = true }
oxc_linter = { workspace = true }
