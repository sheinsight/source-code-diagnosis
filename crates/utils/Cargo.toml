[package]
name = "utils"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
anyhow = "1.0.88"
napi-derive = { workspace = true }
napi = { workspace = true }
rayon = "1.10.0" 
wax = "0.6.0"
ropey = "1.0.0"
pathdiff = {workspace = true}
beans = { path = "../beans" }
log = { workspace = true }
camino = { workspace = true }

oxc = {workspace = true}