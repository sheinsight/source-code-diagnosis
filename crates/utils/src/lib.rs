mod glob;
mod semantic_builder;

use std::{
  fs,
  io::{BufReader, Read},
};

pub use glob::{glob_by, GlobArgs, GlobJsArgs};
pub use semantic_builder::{
  source_type_from_path, SemanticBuilder, SemanticHandler,
};

pub fn read_file_content(path: &std::path::Path) -> anyhow::Result<String> {
  let file = fs::File::open(path)?;
  let mut reader = BufReader::with_capacity(1024 * 1024, file); // 1MB buffer
  let mut content = String::new();
  reader.read_to_string(&mut content)?;
  Ok(content)
}

pub fn is_ts_video(path: &std::path::Path) -> bool {
  if let Ok(mut file) = fs::File::open(path) {
    let mut buffer = [0; 4];
    if file.read_exact(&mut buffer).is_ok() {
      // TS 视频文件的魔数是 0x47
      return buffer[0] == 0x47;
    }
  }
  false
}
