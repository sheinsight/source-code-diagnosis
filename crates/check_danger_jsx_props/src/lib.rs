use std::fmt::Display;

use napi_derive::napi;
use oxc::ast::{AstKind, ast::JSXAttributeItem};
use serde::Serialize;
use utils::{GlobErrorHandler, GlobSuccessHandler};

#[derive(Debug, Clone, Serialize)]
#[napi(object)]
pub struct CheckDangerJsxPropsResponse {
  pub file_path: String,
  pub items: Vec<CheckDangerJsxPropsResponseItem>,
  pub errors: Vec<String>,
}

#[derive(Debug, Clone, Serialize)]
#[napi(object)]
pub struct CheckDangerJsxPropsResponseItem {
  pub name: String,
  pub ast_node: beans::AstNode,
}

impl Display for CheckDangerJsxPropsResponse {
  fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
    write!(f, "{}", serde_json::to_string_pretty(self).unwrap())
  }
}

pub fn check_danger_jsx_props(
  danger_jsx_props: Vec<String>,
  args: utils::GlobArgs,
) -> anyhow::Result<Vec<CheckDangerJsxPropsResponse>> {
  let responses = utils::glob_by_semantic(
    |GlobSuccessHandler {
       semantic,
       relative_path,
       ..
     }| {
      let responses = semantic
        .nodes()
        .into_iter()
        .filter_map(|node| {
          if let AstKind::JSXAttributeItem(jsx_attribute_item) = node.kind() {
            if let JSXAttributeItem::Attribute(jsx_attribute) =
              jsx_attribute_item
            {
              if danger_jsx_props
                .contains(&jsx_attribute.name.get_identifier().name.to_string())
              {
                return Some(CheckDangerJsxPropsResponseItem {
                  name: jsx_attribute.name.get_identifier().name.to_string(),
                  ast_node: beans::AstNode::with_source_and_span(
                    &semantic.source_text(),
                    &jsx_attribute.span,
                  ),
                });
              }
            }
          }
          None
        })
        .collect::<Vec<_>>();
      Some(CheckDangerJsxPropsResponse {
        file_path: relative_path,
        items: responses,
        errors: vec![],
      })
    },
    |GlobErrorHandler {
       relative_path,
       error,
       ..
     }| {
      return Some(CheckDangerJsxPropsResponse {
        file_path: relative_path,
        items: vec![],
        errors: vec![error],
      });
    },
    &args,
  )?;

  Ok(responses)
}
