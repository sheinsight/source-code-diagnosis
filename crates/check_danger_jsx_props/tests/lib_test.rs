use check_danger_jsx_props::check_danger_jsx_props;
use std::env::current_dir;
use utils::{GlobArgs, GlobJsArgs};

#[test]
fn test_check_syntax() -> anyhow::Result<()> {
  let curr = current_dir()?.join("tests").join("features").join("normal");

  let args = GlobArgs {
    cwd: curr.to_string_lossy().to_string(),
    pattern: "*.{js,ts,jsx,tsx}".to_owned(),
    ignore: vec![],
  };

  let mut responses =
    check_danger_jsx_props(vec!["data-if".to_owned()], args).unwrap();

  responses.sort_by(|a, b| a.file_path.cmp(&b.file_path));

  assert_eq!(responses.len(), 2);

  // responses.iter().for_each(|response| {
  //   insta::assert_snapshot!(response.to_string());
  // });

  Ok(())
}

#[test]
fn test_check_syntax_error() -> anyhow::Result<()> {
  let curr = current_dir()?
    .join("tests")
    .join("features")
    .join("syntax_errors");

  let args = GlobJsArgs {
    cwd: curr.to_string_lossy().to_string(),
    pattern: None,
    ignore: None,
  };

  let responses =
    check_danger_jsx_props(vec!["data-if".to_owned()], args.into()).unwrap();

  assert_eq!(responses.len(), 1);

  // insta::assert_snapshot!(responses[0].to_string());

  Ok(())
}
