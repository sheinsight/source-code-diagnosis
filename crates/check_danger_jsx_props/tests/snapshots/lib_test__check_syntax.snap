---
source: crates/check_danger_jsx_props/tests/lib_test.rs
expression: response.to_string()
---
{
  "file_path": "a.jsx",
  "items": [
    {
      "name": "data-if",
      "ast_node": {
        "span": {
          "start": 132,
          "end": 146
        },
        "loc": {
          "start": {
            "line": 6,
            "col": 14
          },
          "end": {
            "line": 6,
            "col": 28
          }
        }
      }
    }
  ],
  "errors": []
}
